### PR TITLE
Correctly close all StatsD clients

### PR DESCRIPTION
**Pull request by marcotc**
The new version of [`dogstatsd-ruby`, v4.9.0,](https://github.com/DataDog/dogstatsd-ruby/releases/tag/v4.9.0) creates a background thread, which needs to be properly closed by calling the previously existing method `#close`.

In a few parts of our tracer and test suite we fail to call `#close` on all instances of the StatsD client. This PR addresses that.

The configuration of Runtime Metric and Health Metrics components reuse the existing `statsd` client from previous configurations, thus directly manipulating the lifecycle of `statsd`. This doesn't seem like a natural relationship, as the `statsd` client is nested inside the Runtime Metrics Worker object, and trying to manipulate its lifecycle from outside creates a few complications. These are documented inline. This also gives us another reason to "flatten" our internal tracer components, in this case the `statsd` client, as that will restore the natural ownership of its lifecycle to the global Components manager.

### CODE REVIEW

**Comment by marcotc** [comment]
`dogstatsd-ruby` `4.9.0` was recently removed, but `5.0.0` will be released shortly with the change changes (just different version number, due to breaking API change).
Even though it's a major release, the breaking changes do not affect `ddtrace`'s statsd calls.

**Comment by marcotc** [comment]
The build failure is due to https://github.com/DataDog/dogstatsd-ruby/pull/175

File: spec/spec_helper.rb
Comment (marcotc) [inline spec/spec_helper.rb:95]: JRuby builds are timing out: https://app.circleci.com/pipelines/github/DataDog/dd-trace-rb/3161/workflows/2c837108-5272-42dc-841a-694d76e30b60/jobs/126109
And build console output is over 400MiB 😢

File: spec/support/tracer_helpers.rb
Comment (marcotc) [inline spec/support/tracer_helpers.rb:164]: Borrowed from https://github.com/DataDog/dd-trace-rb/pull/1421/files#diff-d2e9b89d3ad8e619ce1d915526eef5288a19e54add48fedafed37f18907b872b

File: Gemfile
Comment (marcotc) [inline Gemfile:63]: I don't have push access to `DataDog/dogstatsd-ruby`, so my fork it is for now.

**Comment by marcotc** [comment]
I decided to make this PR pass our build, instead of breaking down all requirements to have a green CI over 3 CI failing PRs.

The first commit contains the main changes. The next two are separate issues that are required for CI to pass. I'm trusting that git history will be enough to help us separate these changes if we need to look back into this changeset.

File: Gemfile
Comment (ivoanjo) [reply Gemfile:63]: This is reasonable, given that it's only used for testing.

File: lib/ddtrace/configuration/components.rb
Comment (ivoanjo) [inline lib/ddtrace/configuration/components.rb:128]: Minor: Is this `enabled = false` needed? Wouldn't stop work even if `enabled` was still `true`?

File: lib/ddtrace/workers/runtime_metrics.rb
Comment (ivoanjo) [inline lib/ddtrace/workers/runtime_metrics.rb:55]: Is this safe? We're calling `.close` on something that is shared with a background thread, so if `Runtime::Metrics` or `Datadog::Statsd` are not thread-safe, this can be problematic.

I was thinking that perhaps we can have an easy fix by moving the close *after* the call to super -- then we know the background activity has stopped, so it should be a safer time to close the resource?

File: spec/ddtrace_integration_spec.rb
Comment (ivoanjo) [inline spec/ddtrace_integration_spec.rb:117]: Minor: Perhaps `to be_truthy` would be a bit more flexible here?

File: spec/spec_helper.rb
Comment (ivoanjo) [reply spec/spec_helper.rb:95]: 400mb of feedback is clearly not helpful. Rather than skipping it entirely in CI, could we count how many errors we've emitted, and after some number, we just stop reporting them, and just include a line at the end of the report with how many errors?

Famously, [valgrind](https://valgrind.org/) does the same:

```
More than 10000000 total errors detected.  I'm not reporting any more.
Final error counts will be inaccurate.  Go fix your program!
```

File: spec/spec_helper.rb
Comment (ivoanjo) [inline spec/spec_helper.rb:189]: I like how this method is like the history of Ruby bugs and features. "Let's just wrap it" turns into "oh but this crashes on old rubies, and we need to use correct delegation on modern rubies...". Ahh computers...

**Review by ivoanjo — COMMENTED** [review]
Thanks for this 🙇 . I'm a bit concerned with the potential race during close, and hence as much as it pains me, I'll hold off on pressing the merge button >_>

File: lib/ddtrace/workers/runtime_metrics.rb
Comment (marcotc) [reply lib/ddtrace/workers/runtime_metrics.rb:55]: I don't have a strong opinion here.
It's rather hard to reason about this worker class, as there are a few layers of pooling/threading/loop that interweave with each other... I always get lost in the process. It seems like something we'd want to "flatten" in the future.

I'll give it a shot and see if works correctly with close after super.

File: lib/ddtrace/configuration/components.rb
Comment (marcotc) [reply lib/ddtrace/configuration/components.rb:128]: It doesn't seem like it's strictly required. This `enabled` is part of the worker pooling/threading/loop pattern, so it's tricky to be sure, but as far as I can reason, it's not actually needed to properly shut down `runtime_metrics` here.

**Review by delner — APPROVED** [review]
Looks good. :+1:

File: spec/spec_helper.rb
Comment (ivoanjo) [inline spec/spec_helper.rb:109]: It may be worth adding a note at the end of the test suite execution too, otherwise if you ignore the output of running a lot of tests (which to be honest, is what I do with such a noisy output as we have right now), you may miss that we still have a bunch of those to fix :)

File: lib/ddtrace/workers/runtime_metrics.rb
Comment (ivoanjo) [reply lib/ddtrace/workers/runtime_metrics.rb]: Yes please! I think the module approach right now makes the combined objects have a lot of behavior and weird cases. I'd really like to move to using composition for these workers, e.g. two objects which communicate with clear APIs, rather than the current approach of having a lot of behavior injected into the class via a combination of include and prepend.

**Pull request merged**
